libc/picolibc: Switch to common malloc implementation

keith-packard opened the pull request:
This series mostly involves adding functionality to the common malloc implementation to make it capable of supporting all of the target requirements provided by the picolibc malloc arena setup code. Those were derived from the newlib functionality originally, but have been simplified to reduce the amount of configuration necessary for each target.

keith-packard committed:
libc/common: Handle MMU/MPU/USERSPACE and alignment requirements in malloc

When CONFIG_COMMON_LIBC_MALLOC_ARENA_SIZE is set to -1, make the malloc
arena use all remaining RAM. When an MMU exists, allocate the arena at
startup using k_mem_map. Otherwise, compute the available memory
automatically and use that.

When an MPU is being used to manage the malloc arena, make sure the heap
respects any MPU alignment requirements. Otherwise, align the heap to
sizeof(double).

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
libc/common: Allocate default malloc arena for non-minimal-libc uses

When using the common malloc implementation on systems not using the
minimal C library, allocate a reasonable default malloc heap according to
the following rules (adopted from the Picolibc heap size defaults):

* When an MMU is available, allocate 16kB.
* When USERSPACE is enabled for a device with an MPU require PoT alignment,
  allocate 1024 bytes
* Otherwise, use all available memory.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
libc/picolibc: Switch to common malloc implementation

Enable the common malloc implementation when using picolibc. Support
existing Picolibc configurations by respecting existing PICOLIBC_HEAP_SIZE
settings.

When PICOLIBC_HEAP_SIZE is set to a value other than -2, then
*always* set COMMON_LIBC_MALLOC_ARENA_SIZE to that value.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
libc/newlib: Remove duplicate declaration of z_malloc_partition

The newlib stanza in libc-hooks.h declares z_malloc_partition, but also
sets Z_MALLOC_PARTITION_EXISTS. That preprocessor symbol controls
whether z_malloc_partition is declared later in the file, causing
a duplicate declaration. Remove the newlib-specific one as redundant.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
The coding guidelines failures are caused by the definition of malloc/calloc/free functions.

cfriedt reviewed (dismissed):
Wow - definitely a creative solution. Thanks for looking at gracefully deprecating old Kconfig options.

stephanosio commented on lib/libc/common/Kconfig:
```diff
@@ -12,14 +12,42 @@ config COMMON_LIBC_MALLOC
 	  API.
 
+config COMMON_LIBC_MALLOC_ARENA_SIZE_DEFAULT
+	int
+	default 0 if MINIMAL_LIBC
+	default 16384 if MMU
+	default -1
+	help
+	  Indicates the default value for COMMON_LIBC_MALLOC_ARENA_SIZE.
+	  This is used when COMMON_LIBC_MALLOC_ARENA_SIZE is set to -2.
+
+	  When using the minimal C library, the default is to have no
+	  malloc heap. Otherwise, on systems with an MMU the default is
+	  16kB and all other systems will default to using all remaining
+	  ram for the malloc heap.
+
 config COMMON_LIBC_MALLOC_ARENA_SIZE
 	int "Size of the common C library malloc arena"
 	depends on COMMON_LIBC_MALLOC
-	default 0
+	default -2
 	help
 	  Indicate the size in bytes of the memory arena used for
 	  common C library malloc() implementation.
 
+	  If set to -2, then the value of
+	  COMMON_LIBC_MALLOC_ARENA_SIZE_DEFAULT will be used instead, and
+	  the remaining rules applied.
+
+	  If set to zero, then no malloc() heap will be available.
+
+	  If set to -1, then all remaining system RAM will be used for this
+	  area.
+
+	  If user mode is enabled, and MPU hardware has requirements that
+	  regions be sized to a power of two and aligned to their size,
+	  then this must be defined as a power of two or a compile error
+	  will result.
```
Having both `COMMON_LIBC_MALLOC_ARENA_SIZE` and `COMMON_LIBC_MALLOC_ARENA_SIZE_DEFAULT` is quite confusing and seems far from ideal to me.

Instead of making `COMMON_LIBC_MALLOC_ARENA_SIZE` function as both the "master switch" for malloc as well as the size of the malloc arena, why not make it specify only the size of the malloc arena, and instead make `COMMON_LIBC_MALLOC` the master switch as it should be?

keith-packard replied:
yeah, as usual it's complicated by the desire to remain compatible with existing application configurations that set `MINIMAL_LIBC_MALLOC_ARENA_SIZE` (or `PICOLIBC_HEAP_SIZE`, once that patch is merged). 

Hrm. I bet this would look better if I changed the default value of the legacy configuration values from `0` to `-1` so I could tell the difference between an application which explicitly defines the heap size as the old default and one which doesn't define the value at all.

I'll take a stab at that and see what it looks like.

keith-packard replied:
To answer the question about "master switch", `malloc` et al. will be available whenever `COMMON_LIBC_MALLOC` is defined, but if the heap is set to zero, then it will never succeed. This makes it work like the minimal C library allocator, but it seems like a user trap to me. Your application would build and run until the first malloc call, at which point it would fail. Should we fix that?

keith-packard replied:
in any case, moving the funky bits for dealing with legacy settings out of the common malloc settings and into the legacy library settings makes things look a lot cleaner. It does mean that the legacy settings will override the common settings, I think that's probably fine, but would appreciate some thought on that.

keith-packard committed:
libc/common: Change how legacy malloc setting works

Change the default value of MINIMAL_LIBC_MALLOC_ARENA_SIZE from 0 to an
otherwise invalid value of -2. This allows the common malloc code to
distinguish between an application which explicitly sets the malloc arena
to zero bytes from an application which does not set the variable at all.

When MINIMAL_LIBC_MALLOC_ARENA_SIZE is set to a value other than -2, then
*always* set COMMON_LIBC_MALLOC_ARENA_SIZE to that value. Emit a compile
warning to let the developer know that their application is using a
deprecated Kconfig variable.

This means that MINIMAL_LIBC_MALLOC_ARENA_SIZE now always overrides
COMMON_LIBC_MALLOC_ARENA_SIZE, but an application setting both will already
be getting a configuration warning for setting the DEPRECATED value
anyways, which should serve as an adequate indication to the user.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
boards/qemu_x86: Switch reserved frames count to common malloc

With picolibc switching to the common malloc implementation, the
computation of how many page frames to reserve needs to change as
well. Note that this will now also reserve page frames when the minimal C
library enables the allocator.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard committed:
tests,samples: Fix picolibc malloc arena settings

With picolibc moving to using the common malloc implementation, samples and
tests with picolibc-specific settings need to switch to using the common
malloc settings instead.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
The coding guidlines check failed because this patch touches malloc.c which defines 'reserved' identifiers. Would it be reasonable to allow reserved identifiers in lib/libc files?

keith-packard commented:
> Wow - definitely a creative solution. Thanks for looking at gracefully deprecating old Kconfig options.

You might want to re-review that after @stephanosio encouraged me to refactor the configuration bits. With this change, the old Kconfig options now take priority over the new ones. I *think* the overall result is cleaner, but would like know if there are concerns about that particular behavior.

stephanosio commented on include/zephyr/sys/libc-hooks.h:
```diff
@@ -41,13 +41,12 @@ __syscall size_t zephyr_fwrite(const void *ZRESTRICT ptr, size_t size,
-#  ifdef CONFIG_MINIMAL_LIBC
-#   define CONFIG_COMMON_LIBC_MALLOC_ARENA_SIZE CONFIG_MINIMAL_LIBC_MALLOC_ARENA_SIZE
-#  else
-#   define CONFIG_COMMON_LIBC_MALLOC_ARENA_SIZE 0
-#  endif
+#  define CONFIG_COMMON_LIBC_MALLOC_ARENA_SIZE CONFIG_MINIMAL_LIBC_MALLOC_ARENA_SIZE
+# elif defined(CONFIG_PICOLIBC) && (CONFIG_PICOLIBC_HEAP_SIZE != -2)
```
> With this change, the old Kconfig options now take priority over the new ones

Since setting these deprecated Kconfigs do not print a deprecation warning (because they are `int` and cannot select `DEPRECATED`), I think it would help if we print a warning if these deprecated Kconfigs are set to a non-default value.

Something like:

```
# if defined(CONFIG_MINIMAL_LIBC) && (CONFIG_MINIMAL_LIBC_MALLOC_ARENA_SIZE != -2)
#    warning ...
...
```

keith-packard replied:
yeah, that's a good idea -- will also trigger compile errors when built with -Werror

stephanosio reviewed (commented):
Looks much better now, thanks!

Just a minor comment about the deprecation warning.

stephanosio commented:
It looks like we still have a few leftovers using the deprecated configs:

```
ccache /opt/toolchains/zephyr-sdk-0.16.0/aarch64-zephyr-elf/bin/aarch64-zephyr-elf-gcc -DKERNEL -D__ZEPHYR__=1 -I/__w/zephyr/zephyr/include -I/__w/zephyr/zephyr/twister-out/qemu_cortex_a53/samples/drivers/virtualization/ivshmem/doorbell/sample.virtualization.ivshmem.doorbell.arm64/zephyr/include/generated -I/__w/zephyr/zephyr/soc/arm64/qemu_cortex_a53 -isystem /__w/zephyr/zephyr/lib/libc/minimal/include -isystem /opt/toolchains/zephyr-sdk-0.16.0/aarch64-zephyr-elf/bin/../lib/gcc/aarch64-zephyr-elf/12.2.0/include -isystem /opt/toolchains/zephyr-sdk-0.16.0/aarch64-zephyr-elf/bin/../lib/gcc/aarch64-zephyr-elf/12.2.0/include-fixed -fno-strict-aliasing -Werror -Os -imacros /__w/zephyr/zephyr/twister-out/qemu_cortex_a53/samples/drivers/virtualization/ivshmem/doorbell/sample.virtualization.ivshmem.doorbell.arm64/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-a53 -mabi=lp64 --sysroot=/opt/toolchains/zephyr-sdk-0.16.0/aarch64-zephyr-elf/aarch64-zephyr-elf -imacros /__w/zephyr/zephyr/include/zephyr/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-unused-but-set-variable -Werror=implicit-int -fno-pic -fno-pie -fno-asynchronous-unwind-tables -fno-reorder-functions --param=min-pagesize=0 -fno-defer-pop -fmacro-prefix-map=/__w/zephyr/zephyr/samples/drivers/virtualization/ivshmem/doorbell=CMAKE_SOURCE_DIR -fmacro-prefix-map=/__w/zephyr/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/__w/zephyr=WEST_TOPDIR -ffunction-sections -fdata-sections -mno-outline-atomics -moverride=tune=no_ldp_stp_qregs -std=c99 -nostdinc -MD -MT CMakeFiles/app.dir/src/ivshmem.c.obj -MF CMakeFiles/app.dir/src/ivshmem.c.obj.d -o CMakeFiles/app.dir/src/ivshmem.c.obj -c /__w/zephyr/zephyr/samples/drivers/virtualization/ivshmem/doorbell/src/ivshmem.c
In file included from /__w/zephyr/zephyr/samples/drivers/virtualization/ivshmem/doorbell/src/ivshmem.c:16:
/__w/zephyr/zephyr/include/zephyr/sys/libc-hooks.h:47:4: error: #warning Using deprecated setting CONFIG_MINIMAL_LIBC_MALLOC_ARENA_SIZE [-Werror=cpp]
   47 | #  warning Using deprecated setting CONFIG_MINIMAL_LIBC_MALLOC_ARENA_SIZE
      |    ^~~~~~~
cc1: all warnings being treated as errors
```

https://github.com/search?q=repo%3Azephyrproject-rtos%2Fzephyr%20MINIMAL_LIBC_MALLOC_ARENA_SIZE&type=code

stephanosio committed:
boards, samples: Fix minimal libc malloc arena settings

With the minimal libc moving to using the common malloc implementation,
boards and samples with minimal libc-specific settings need to switch
to using the common malloc settings instead.

Signed-off-by: Stephanos Ioannidis <stephanos.ioannidis@nordicsemi.no>

stephanosio commented:
> It looks like we still have a few leftovers using the deprecated configs:

I have pushed a commit to fix that.

keith-packard commented:
> I have pushed a commit to fix that.

Thanks! Afraid I didn't get a chance to come back and check on the progress of the CI run. And, we've also had a chance to verify that the warnings are working.

keith-packard commented:
> I have pushed a commit to fix that.

Do you want me to reorder the commits so this fix lands before the warnings are enabled? Or are bisect failures caused by things like this OK?

keith-packard committed:
libc/picolibc: Warn when apps use PICOLIBC_HEAP_SIZE setting

Now that the Zephyr tree has had uses of this variable removed, emit a
compile warning to let the developer know that their application is using a
deprecated Kconfig variable.

Signed-off-by: Keith Packard <keithp@keithp.com>

keith-packard commented:
> We do not strictly require bisectability on every commit; but, in this case, it would not hurt to properly order the commits for bisectability since doing so does not require much effort.

Seems reasonable. I've moved your fix earlier before the warning was added and then split the picolibc change into two pieces sandwiching the fixes for `PICOLIBC_HEAP_SIZE`.

stephanosio commented:
@cfriedt Could you take another look?

cfriedt commented on lib/libc/common/Kconfig:
```diff
@@ -15,11 +15,28 @@ config COMMON_LIBC_MALLOC
 	depends on COMMON_LIBC_MALLOC
-	default 0
+	default 0 if MINIMAL_LIBC
+	default 16384 if MMU
```
Does 16384 work for all arches with an MMU?

keith-packard replied:
This is inherited from the picolibc configuration, with the goal of not changing anything when using that. However, I note that picolibc also has `default 1024 if USERSPACE && MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT`, which I didn't manage to copy over.

A couple of options here:

 1. Merge as-is, expecting people using a malloc heap will end up setting the size manually most of the time anyways.
 2. Do a test with the 'picolibc by default' and this branch merged together and see what breaks -- that would catch configurations expecting the above condition to be used.
 3. Add the above condition so that picolibc users don't see any change.

I'm leaning towards option 3. as that introduces the least change for existing picolibc users.

keith-packard replied:
Yeah, I thought that's what I had done before, but forgot that I removed the missing default clause when I had improved the common code to deal with MPU POT alignment requirements so that the missing default wasn't necessary anymore. We can revisit what the right defaults are when we've got more experience using this in more applications.

cfriedt reviewed (dismissed):
LGTM. Just a minor question about 16384, but not blocking

stephanosio commented:
> Error: lib/libc/common/source/stdlib/malloc.c:113:WARNING: Violation to rule 21.2 (Should not used a reserved identifier) - calloc	
WARNING: Violation to rule 21.2 (Should not used a reserved identifier) - free	
WARNING: Violation to rule 21.2 (Should not used a reserved identifier) - malloc	
WARNING: Violation to rule 21.2 (Should not used a reserved identifier) - realloc

Coding guidelines check failures are false positives. Merging.
